Add contributor guidelines page (#943)
opensafely/documentation: docs/contributing.md

# Contributing

OpenSAFELY is an open source project; all our code is available for inspection and reuse, and we can accept contributions in the following ways:

## Documentation

Typos and other small improvements are always very welcome. Raise an issue, or even better a Pull Request, in the [documentation repo](https://github.com/opensafely/documentation/). We will incorporate these contributions after we've checked them for consistency with our house style.

We welcome proposals for longer topic-based guides. Suggest them as an [issue in the documentation repo](https://github.com/opensafely/documentation/issues), and we'll work with you to define an overall structure and format, before you write and submit the draft.

## Reusable actions

> These are units of software that solve a problem for several studies without the need to copy-and-paste between them, [described in more detail here](./actions-reusable.md).  They can be shared between researchers, even between groups that use different programming languages, and are one of the best ways you can make contributions that benefit the community.  Because they have a clear, pluggable interface with the rest of OpenSAFELY, it's easy for us to accept new reusable actions into our [actions library](https://actions.opensafely.org/). If you've written a reusable action you'd like to contribute to the actions library, please get in touch at [team@opensafely.org](mailto:team@opensafely.org).

Reviewer: ```suggestion
These are units of software that solve a problem for several studies without the need to copy-and-paste between them, [described in more detail here](./actions-reusable.md).  They can be shared between researchers, even between groups that use different programming languages, and are one of the best ways you can make contributions that benefit the community. If you've written a reusable action you'd like to contribute to the actions library, please get in touch at [team@opensafely.org](mailto:team@opensafely.org).
```

I don't think we should accept anything as a reusable action, so would emphasize getting in touch over how "easy" it is for us to accept new reusable actions.

Reviewer: I'll add an issue to the data team backlog for documenting reusable action "best practice".

## Peer support

We encourage researchers to post questions in the [Q&A Forum](https://github.com/opensafely/documentation/discussions). We would love more people to chip in and attempt to answer questions!

## Core code
We would like to accept _ad hoc_ contributions to the [core code](https://github.com/opensafely-core/), but this is relatively hard for us to do. New features and bugfixes need a lot work from our side to integrate and maintain, because a lot of the design and prioritisation decisions are made with lots of different parts of the framework in mind. All but the most trivial changes also come with intellectual property considerations, so we also need to secure Contributor Agreements from individuals or their employers.  We plan to seek funding to help resource this kind of work, as we believe it can really strengthen and build our community, but for the time being, we suggest [filing bug reports or feature requests](https://github.com/opensafely-core) rather than contributing code.

> 

Reviewer 2: > All but the most trivial changes also come with intellectual property considerations, so we also need to secure Contributor Agreements from individuals or their employers.

I wonder about this line.  It doesn't sound very open source to be talking about intellectual property.  Does it need to be mentioned at all?

Reviewer 3: I don't think it hurts to mention the topic - as the paragraph at the moment is just giving a few template justifications why we might need to ignore an "easy to merge" external patch. I think you have a point about avoiding the buzzphrase "intellectual property" though, as some people might bridle at it - maybe something more vague like "...also come with legal risks...".

Reviewer 4: You could also switch out:

> All but the most trivial changes also come with intellectual property considerations…

for

> Most code changes have an associated administrative and legal burden: we need to secure contributor agreements from individual or their employers.

> An exception is where external contributors are in a position to dedicate significant resource (e.g. more than one full-time person equivalent) for a prolonged period of time. In these cases, we will be happy to consider allocating resource from our side to help with the planning and integration work; just get in touch with your proposed features at [team@opensafely.org](mailto:team@opensafely.org).

Reviewer 5: > consider allocating resource

Very small point, but I try to avoid this word when referring to people. Could resource be changed to "someone" or "a person" instead?

Reviewer 4: Here's my go at a slight rewrite for conciseness:

```suggestion
Where external contributors can dedicate significant time and effort (e.g. more than one full-time person equivalent), we will be happy to consider helping you with the planning and integration work. Get in touch with your proposed features at [team@opensafely.org](mailto:team@opensafely.org).
```
